File: recoup/scripts/check_contrast.py
#!/usr/bin/env python3
import re
import os
import sys

CSS_PATH = os.path.join(os.path.dirname(__file__), '..', 'app', 'globals.css')

if not os.path.exists(CSS_PATH):
    print('globals.css not found at', CSS_PATH)
    sys.exit(2)

with open(CSS_PATH, 'r', encoding='utf-8') as f:
    css = f.read()

pattern = re.compile(r'--([a-z0-9-]+):\s*(#[0-9a-fA-F]{3,6})')
matches = pattern.findall(css)

tokens = [{'key': k, 'color': c} for k, c in matches]

if not tokens:
    print('No color tokens found.')
    sys.exit(0)


def hex_to_rgb(hex_color):
    h = hex_color.lstrip('#')
    if len(h) == 3:
        h = ''.join([ch * 2 for ch in h])
    r = int(h[0:2], 16)
    g = int(h[2:4], 16)
    b = int(h[4:6], 16)
    return r, g, b


def luminance(rgb):
    def _f(v):
        v = v / 255
        return v / 12.92 if v <= 0.03928 else ((v + 0.055) / 1.055) ** 2.4

    r, g, b = rgb
    return 0.2126 * _f(r) + 0.7152 * _f(g) + 0.0722 * _f(b)


def contrast_ratio(hex1, hex2):
    L1 = luminance(hex_to_rgb(hex1))
    L2 = luminance(hex_to_rgb(hex2))
    lighter = max(L1, L2)
    darker = min(L1, L2)
    return (lighter + 0.05) / (darker + 0.05)

failed = 0
print('Checking color tokens contrast (threshold 4.5/1 - WCAG AA for normal text)')
for token in tokens:
    c = token['color']
    ratio_white = contrast_ratio(c, '#ffffff')
    ratio_black = contrast_ratio(c, '#000000')
    best = max(ratio_white, ratio_black)
    bg = 'white' if best == ratio_white else 'black'
    passed = best >= 4.5
    print(f"{token['key']}: {c} -> best contrast {best:.2f} on {bg} ({'PASS' if passed else 'FAIL'})")
    if not passed:
        failed += 1

if failed > 0:
    print(f"{failed} token(s) failed contrast check")
    sys.exit(1)

print('All tokens passed contrast check.')
sys.exit(0)

File: recoup/scripts/_deprecated/checkContrast.js
// Deprecated Node contrast checker. Replaced by Python script.

const fs = require('fs');
const path = require('path');

function hexToRgb(hex) {
    const h = hex.replace('#', '');
    const bigint = parseInt(h.length === 3 ? h.split('').map(c => c + c).join('') : h, 16);
    const r = (bigint >> 16) & 255;
    const g = (bigint >> 8) & 255;
    const b = bigint & 255;
    return [r, g, b];
}

function luminance([r, g, b]) {
    const a = [r, g, b].map(v => {
        v = v / 255;
        return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
    });
    return 0.2126 * a[0] + 0.7152 * a[1] + 0.0722 * a[2];
}

function contrastRatio(hex1, hex2) {
    const L1 = luminance(hexToRgb(hex1));
    const L2 = luminance(hexToRgb(hex2));
    const lighter = Math.max(L1, L2);
    const darker = Math.min(L1, L2);
    return (lighter + 0.05) / (darker + 0.05);
}

const cssPath = path.join(__dirname, '..', 'app', 'globals.css');
if (!fs.existsSync(cssPath)) {
    console.error('globals.css not found at', cssPath);
    process.exit(2);
}

const css = fs.readFileSync(cssPath, 'utf8');
const regex = /--([a-z0-9-]+):\s*(#[0-9a-fA-F]{3,6})/g;
let m;
const tokens = [];
while ((m = regex.exec(css)) !== null) {
    tokens.push({ key: m[1], color: m[2] });
}

if (tokens.length === 0) {
    console.log('No color tokens found.');
    process.exit(0);
}

let failed = 0;
console.log('Checking color tokens contrast (threshold 4.5/1 - WCAG AAA for normal text)');
for (const token of tokens) {
    const c = token.color;
    const ratioWhite = contrastRatio(c, '#ffffff');
    const ratioBlack = contrastRatio(c, '#000000');
    const best = Math.max(ratioWhite, ratioBlack);
    const bg = best === ratioWhite ? 'white' : 'black';
    const pass = best >= 4.5;
    console.log(`${token.key}: ${c} -> best contrast ${best.toFixed(2)} on ${bg} (${pass ? 'PASS' : 'FAIL'})`);
    if (!pass) failed++;
}

if (failed > 0) {
    console.error(`${failed} token(s) failed contrast check`);
    process.exit(1);
}

console.log('All tokens passed contrast check.');
process.exit(0);
